File: day01/main.hs
import System.IO;

import Data.List;
import Data.List.Split;
import Data.Maybe;
import Text.Printf;

parenToOffset :: Char -> Int
parenToOffset c
    | c == '(' = 1
    | c == ')' = -1
    | otherwise = 0

toCurrentLevelImpl :: ([Int], [Int]) -> ([Int], [Int])
toCurrentLevelImpl (arr, []) = toCurrentLevelImpl (arr, [0])
toCurrentLevelImpl ([], a) = ([], a)
toCurrentLevelImpl (instructions, state) =
    toCurrentLevelImpl (tail instructions, [(head state + head instructions)] ++ state)

toCurrentLevel x = reverse $ snd (toCurrentLevelImpl (x, []))

levelsAtIndex contents = toCurrentLevel (map parenToOffset contents)

main = do
    contents <- readFile "data.txt"
    printf "Task 1: %d\n" (last $ levelsAtIndex contents) 
    printf "Task 2: %d\n" (fromMaybe (-1) (elemIndex (-1) (levelsAtIndex contents)))

File: day01/main.py
data = [0] + list(map(lambda c: 1 if c == '(' else -1 if c == ')' else 0, open('data.txt', 'r').read()))

temp = 0;
for i, v in enumerate(data[1:]):
	data[i+1] = data[i] + v

print('Task 1: %d' % data[-1])
print('Task 1: %d' % data.index(-1))
